File: android/app/src/main/java/sbtbiswas/AidenOnTheGo/features/bots/AidenBotImagePlaygroundView.kt
package sbtbiswas.AidenOnTheGo.features.bots

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AutoAwesome
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Info
import androidx.compose.material.icons.filled.PhotoLibrary
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import sbtbiswas.AidenOnTheGo.ui.theme.AidenTheme
import java.io.File
import java.io.FileOutputStream
import java.util.UUID

data class AidenBotImagePlaygroundIdentity(
    val name: String,
    val purpose: String
) {
    val conceptTexts: List<String>
        get() = listOf(name.trim().take(80), purpose.trim().take(240)).filter { it.isNotEmpty() }
}

enum class AidenBotImagePlaygroundFallbackReason(
    val title: String,
    val message: String
) {
    UNSUPPORTED(
        title = "Image Playground isn't supported on this device",
        message = "You can keep using the semantic avatar on this device."
    ),
    RESTRICTED(
        title = "Image creation is restricted",
        message = "Image Playground may be restricted by device settings. You can keep using the semantic avatar."
    ),
    MODEL_UNAVAILABLE(
        title = "Apple's image model isn't ready",
        message = "Apple's image model may still be downloading or may be unavailable. Try again later, or use the semantic avatar."
    ),
    USAGE_LIMIT(
        title = "Image creation is temporarily limited",
        message = "Apple's image creation limit may have been reached. Try again later, or use the semantic avatar."
    ),
    UPDATE_REQUIRED(
        title = "Update to create a Bot image",
        message = "Aiden needs newer OS support to limit Image Playground to non-personalized styles. You can keep using the semantic avatar."
    ),
    SYSTEM_UNAVAILABLE(
        title = "Image Playground isn't available",
        message = "Apple doesn't currently make Image Playground available on this device. You can keep using the semantic avatar."
    ),
    CANDIDATE_COPY_FAILED(
        title = "That image couldn't be prepared",
        message = "The selected image couldn't be copied into Aiden safely. Choose another image, or use the semantic avatar."
    )
}

sealed class AidenBotImagePlaygroundPresentationPhase {
    object READY : AidenBotImagePlaygroundPresentationPhase()
    object PRESENTING : AidenBotImagePlaygroundPresentationPhase()
    object CANCELLED : AidenBotImagePlaygroundPresentationPhase()
    object ACCEPTED : AidenBotImagePlaygroundPresentationPhase()
    data class FALLBACK(val reason: AidenBotImagePlaygroundFallbackReason) : AidenBotImagePlaygroundPresentationPhase()
}

class AidenBotImagePlaygroundPresentationState(
    initialPhase: AidenBotImagePlaygroundPresentationPhase = AidenBotImagePlaygroundPresentationPhase.READY
) {
    var phase by mutableStateOf(initialPhase)
        private set

    fun requestPresentation(systemAvailable: Boolean) {
        phase = if (systemAvailable) AidenBotImagePlaygroundPresentationPhase.PRESENTING
        else AidenBotImagePlaygroundPresentationPhase.FALLBACK(AidenBotImagePlaygroundFallbackReason.SYSTEM_UNAVAILABLE)
    }

    fun cancel() {
        phase = AidenBotImagePlaygroundPresentationPhase.CANCELLED
    }

    fun acceptCopiedCandidate() {
        phase = AidenBotImagePlaygroundPresentationPhase.ACCEPTED
    }

    fun failCandidateCopy() {
        phase = AidenBotImagePlaygroundPresentationPhase.FALLBACK(AidenBotImagePlaygroundFallbackReason.CANDIDATE_COPY_FAILED)
    }

    fun showFallback(reason: AidenBotImagePlaygroundFallbackReason) {
        phase = AidenBotImagePlaygroundPresentationPhase.FALLBACK(reason)
    }
}

class AidenBotImagePlaygroundCandidateStore(
    private val directory: File = File(System.getProperty("java.io.tmpdir"), "AidenBotImageCandidates").apply { mkdirs() }
) {
    companion object {
        const val MAX_SOURCE_BYTES = 32 * 1024 * 1024
        const val MAX_RETAINED_CANDIDATES = 8
        const val STALE_CANDIDATE_AGE_MS = 24 * 60 * 60 * 1000L
    }

    fun copyImmediately(sourceFile: File): File {
        if (!sourceFile.exists() || !sourceFile.isFile || sourceFile.length() <= 0) {
            throw IllegalArgumentException("Invalid source file")
        }
        if (sourceFile.length() > MAX_SOURCE_BYTES) {
            throw IllegalArgumentException("Source file too large")
        }

        pruneOwnedCandidates(retainingAtMost = MAX_RETAINED_CANDIDATES - 1)
        val destination = File(directory, "candidate-${UUID.randomUUID()}.image")
        sourceFile.copyTo(destination, overwrite = true)
        return destination
    }

    fun copyImmediately(data: ByteArray): File {
        if (data.isEmpty() || data.size > MAX_SOURCE_BYTES) {
            throw IllegalArgumentException("Invalid candidate data")
        }
        pruneOwnedCandidates(retainingAtMost = MAX_RETAINED_CANDIDATES - 1)
        val destination = File(directory, "candidate-${UUID.randomUUID()}.image")
        FileOutputStream(destination).use { it.write(data) }
        return destination
    }

    fun removeOwnedCandidate(file: File) {
        if (isOwnedCandidate(file)) {
            file.delete()
        }
    }

    fun removeAllOwnedCandidates() {
        val files = directory.listFiles() ?: return
        for (file in files) {
            if (isOwnedCandidate(file)) {
                file.delete()
            }
        }
    }

    fun pruneOwnedCandidates(
        now: Long = System.currentTimeMillis(),
        retainingAtMost: Int = MAX_RETAINED_CANDIDATES
    ) {
        val files = directory.listFiles { file -> isOwnedCandidate(file) } ?: return
        val sorted = files.sortedByDescending { it.lastModified() }
        for (index in sorted.indices) {
            val file = sorted[index]
            if (index >= retainingAtMost || (now - file.lastModified()) > STALE_CANDIDATE_AGE_MS) {
                file.delete()
            }
        }
    }

    private fun isOwnedCandidate(file: File): Boolean {
        return file.parentFile?.absolutePath == directory.absolutePath &&
                file.name.startsWith("candidate-") &&
                file.name.endsWith(".image")
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun AidenBotImagePlaygroundSheet(
    botName: String,
    botPurpose: String = "",
    onDismiss: () -> Unit,
    onImageSelected: (ByteArray) -> Unit
) {
    val palette = AidenTheme.palette
    var prompt by remember { mutableStateOf(if (botPurpose.isNotEmpty()) botPurpose else "A friendly AI assistant avatar named $botName") }
    var presentationState = remember { AidenBotImagePlaygroundPresentationState() }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(24.dp)
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier.fillMaxWidth()
        ) {
            Icon(
                imageVector = Icons.Default.AutoAwesome,
                contentDescription = null,
                tint = palette.accent,
                modifier = Modifier.size(28.dp)
            )
            Spacer(modifier = Modifier.width(12.dp))
            Text(
                text = "Avatar Studio",
                style = MaterialTheme.typography.titleLarge,
                fontWeight = FontWeight.Bold,
                color = palette.foreground,
                modifier = Modifier.weight(1f)
            )
            IconButton(onClick = onDismiss) {
                Icon(Icons.Default.Close, contentDescription = "Close", tint = palette.foreground)
            }
        }

        Spacer(modifier = Modifier.height(16.dp))

        // Fallback explanation card for Android platform
        Card(
            modifier = Modifier.fillMaxWidth(),
            colors = CardDefaults.cardColors(containerColor = palette.raised),
            shape = RoundedCornerShape(12.dp)
        ) {
            Column(modifier = Modifier.padding(16.dp)) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Icon(Icons.Default.Info, contentDescription = null, tint = palette.secondary, modifier = Modifier.size(20.dp))
                    Spacer(modifier = Modifier.width(8.dp))
                    Text(
                        text = "Image Generation",
                        style = MaterialTheme.typography.titleSmall,
                        fontWeight = FontWeight.Bold,
                        color = palette.foreground
                    )
                }
                Spacer(modifier = Modifier.height(8.dp))
                Text(
                    text = "Bot photos generated on macOS can be synchronized to Android. You can also customize your Bot with the built-in Semantic Avatar studio.",
                    style = MaterialTheme.typography.bodySmall,
                    color = palette.secondary
                )
            }
        }

        Spacer(modifier = Modifier.height(16.dp))

        OutlinedTextField(
            value = prompt,
            onValueChange = { prompt = it },
            label = { Text("Avatar Description") },
            placeholder = { Text("Describe the appearance of your bot...") },
            minLines = 3,
            maxLines = 5,
            shape = RoundedCornerShape(12.dp),
            modifier = Modifier.fillMaxWidth()
        )

        Spacer(modifier = Modifier.height(24.dp))

        Button(
            onClick = onDismiss,
            colors = ButtonDefaults.buttonColors(containerColor = palette.accent),
            shape = RoundedCornerShape(12.dp),
            modifier = Modifier.fillMaxWidth()
        ) {
            Text("Done", color = Color.White, fontWeight = FontWeight.Bold)
        }
    }
}

File: ios/AidenOnTheGo/Features/Bots/AidenBotImagePlaygroundView.swift
import Foundation
import ImagePlayground
import SwiftUI

struct AidenBotImagePlaygroundIdentity: Equatable, Sendable {
    let name: String
    let purpose: String

    init(name: String, purpose: String) {
        self.name = Self.visibleText(name, maximumCharacters: 80)
        self.purpose = Self.visibleText(purpose, maximumCharacters: 240)
    }

    var conceptTexts: [String] {
        [name, purpose].filter { !$0.isEmpty }
    }

    private static func visibleText(_ value: String, maximumCharacters: Int) -> String {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return String(trimmed.prefix(maximumCharacters))
    }
}

enum AidenBotImagePlaygroundFallbackReason: Equatable, Sendable {
    // Use a specific case only when the caller directly knows that fact.
    // `supportsImagePlayground == false` always maps to `systemUnavailable`;
    // Aiden does not infer Apple's restriction, model, or usage-limit state.
    case unsupported
    case restricted
    case modelUnavailable
    case usageLimit
    case updateRequired
    case systemUnavailable
    case candidateCopyFailed

    var title: String {
        switch self {
        case .unsupported:
            "Image Playground isn't supported on this device"
        case .restricted:
            "Image creation is restricted"
        case .modelUnavailable:
            "Apple's image model isn't ready"
        case .usageLimit:
            "Image creation is temporarily limited"
        case .updateRequired:
            "Update to create a Bot image"
        case .systemUnavailable:
            "Image Playground isn't available"
        case .candidateCopyFailed:
            "That image couldn't be prepared"
        }
    }

    var message: String {
        switch self {
        case .unsupported:
            "You can keep using the semantic avatar on this device."
        case .restricted:
            "Image Playground may be restricted by device settings. You can keep using the semantic avatar."
        case .modelUnavailable:
            "Apple's image model may still be downloading or may be unavailable. Try again later, or use the semantic avatar."
        case .usageLimit:
            "Apple's image creation limit may have been reached. Try again later, or use the semantic avatar."
        case .updateRequired:
            "Aiden needs iOS or iPadOS 18.4 or later to limit Image Playground to Apple's non-personalized styles. You can keep using the semantic avatar."
        case .systemUnavailable:
            "Apple doesn't currently make Image Playground available on this device. You can keep using the semantic avatar."
        case .candidateCopyFailed:
            "The selected image couldn't be copied into Aiden safely. Choose another image, or use the semantic avatar."
        }
    }
}

struct AidenBotImagePlaygroundPresentationState: Equatable, Sendable {
    enum Phase: Equatable, Sendable {
        case ready
        case presenting
        case cancelled
        case accepted
        case fallback(AidenBotImagePlaygroundFallbackReason)
    }

    private(set) var phase: Phase = .ready

    mutating func requestPresentation(systemAvailable: Bool) {
        phase = systemAvailable ? .presenting : .fallback(.systemUnavailable)
    }

    mutating func cancel() {
        phase = .cancelled
    }

    mutating func acceptCopiedCandidate() {
        phase = .accepted
    }

    mutating func failCandidateCopy() {
        phase = .fallback(.candidateCopyFailed)
    }

    mutating func showFallback(_ reason: AidenBotImagePlaygroundFallbackReason) {
        phase = .fallback(reason)
    }
}

enum AidenBotImagePlaygroundCandidateCopyError: Error, Equatable {
    case invalidSource
    case sourceTooLarge
    case copyFailed
}

/// Copies the system-owned completion URL while it is still valid. The copied
/// file is an ephemeral candidate; normalization and paired-Mac upload own its
/// later lifecycle.
struct AidenBotImagePlaygroundCandidateStore {
    // The system result is bounded before decode, then the lifecycle
    // normalizer enforces the tighter 4 MiB canonical upload bound.
    static let maximumSourceBytes = 32 * 1_048_576
    static let maximumRetainedCandidates = 8
    static let staleCandidateAge: TimeInterval = 24 * 60 * 60

    let directory: URL

    init(directory: URL = FileManager.default.temporaryDirectory
        .appending(path: "AidenBotImageCandidates", directoryHint: .isDirectory)) {
        self.directory = directory
    }

    func copyImmediately(fromSystemCompletionURL sourceURL: URL) throws -> URL {
        guard sourceURL.isFileURL else {
            throw AidenBotImagePlaygroundCandidateCopyError.invalidSource
        }
        defer { try? FileManager.default.removeItem(at: sourceURL) }

        let values: URLResourceValues
        do {
            values = try sourceURL.resourceValues(
                forKeys: [.isRegularFileKey, .isSymbolicLinkKey, .fileSizeKey]
            )
        } catch {
            throw AidenBotImagePlaygroundCandidateCopyError.invalidSource
        }
        guard values.isRegularFile == true,
              values.isSymbolicLink != true,
              let byteCount = values.fileSize,
              byteCount > 0 else {
            throw AidenBotImagePlaygroundCandidateCopyError.invalidSource
        }
        guard byteCount <= Self.maximumSourceBytes else {
            throw AidenBotImagePlaygroundCandidateCopyError.sourceTooLarge
        }

        let fileManager = FileManager.default
        let destination = directory.appending(path: "candidate-\(UUID().uuidString).image")
        do {
            try fileManager.createDirectory(
                at: directory,
                withIntermediateDirectories: true,
                attributes: [
                    .posixPermissions: 0o700,
                    .protectionKey: FileProtectionType.complete,
                ]
            )
            try pruneOwnedCandidates(now: Date(), retainingAtMost: Self.maximumRetainedCandidates - 1)
            try fileManager.copyItem(at: sourceURL, to: destination)
            let copiedValues = try destination.resourceValues(
                forKeys: [.isRegularFileKey, .isSymbolicLinkKey, .fileSizeKey]
            )
            guard copiedValues.isRegularFile == true,
                  copiedValues.isSymbolicLink != true,
                  let copiedByteCount = copiedValues.fileSize,
                  (1...Self.maximumSourceBytes).contains(copiedByteCount) else {
                throw AidenBotImagePlaygroundCandidateCopyError.invalidSource
            }
            try fileManager.setAttributes(
                [
                    .posixPermissions: 0o600,
                    .protectionKey: FileProtectionType.complete,
                ],
                ofItemAtPath: destination.path
            )
            var resourceValues = URLResourceValues()
            resourceValues.isExcludedFromBackup = true
            var mutableDestination = destination
            try mutableDestination.setResourceValues(resourceValues)
            return destination
        } catch {
            try? fileManager.removeItem(at: destination)
            throw AidenBotImagePlaygroundCandidateCopyError.copyFailed
        }
    }

    func removeOwnedCandidate(at candidateURL: URL) {
        guard Self.isOwnedCandidate(candidateURL, in: directory) else { return }
        try? FileManager.default.removeItem(at: candidateURL)
    }

    func removeAllOwnedCandidates() {
        guard let candidates = try? ownedCandidates() else { return }
        for candidate in candidates {
            try? FileManager.default.removeItem(at: candidate.url)
        }
    }

    func pruneOwnedCandidates(now: Date = Date()) {
        try? pruneOwnedCandidates(now: now, retainingAtMost: Self.maximumRetainedCandidates)
    }

    private func pruneOwnedCandidates(now: Date, retainingAtMost maximumCount: Int) throws {
        let candidates = try ownedCandidates().sorted { $0.modifiedAt > $1.modifiedAt }
        for (index, candidate) in candidates.enumerated()
        where index >= maximumCount || now.timeIntervalSince(candidate.modifiedAt) > Self.staleCandidateAge {
            try? FileManager.default.removeItem(at: candidate.url)
        }
    }

    private func ownedCandidates() throws -> [(url: URL, modifiedAt: Date)] {
        let fileManager = FileManager.default
        guard fileManager.fileExists(atPath: directory.path) else { return [] }
        let keys: Set<URLResourceKey> = [.isRegularFileKey, .contentModificationDateKey]
        return try fileManager.contentsOfDirectory(
            at: directory,
            includingPropertiesForKeys: Array(keys),
            options: [.skipsHiddenFiles]
        ).compactMap { url in
            guard Self.isOwnedCandidate(url, in: directory) else { return nil }
            let values = try url.resourceValues(forKeys: keys)
            guard values.isRegularFile == true else { return nil }
            return (url, values.contentModificationDate ?? .distantPast)
        }
    }

    private static func isOwnedCandidate(_ url: URL, in directory: URL) -> Bool {
        let name = url.lastPathComponent
        return isDescendant(url, of: directory)
            && name.hasPrefix("candidate-")
            && name.hasSuffix(".image")
    }

    private static func isDescendant(_ url: URL, of directory: URL) -> Bool {
        let candidateComponents = url.standardizedFileURL.pathComponents
        let directoryComponents = directory.standardizedFileURL.pathComponents
        guard candidateComponents.count > directoryComponents.count else { return false }
        return Array(candidateComponents.prefix(directoryComponents.count)) == directoryComponents
    }
}

func aidenBotImagePlaygroundCleanupAfterProcessLaunch(
    candidateStore: AidenBotImagePlaygroundCandidateStore = .init()
) {
    // No accepted candidate survives in memory across a process launch, so
    // every app-owned temporary candidate is crash residue at this boundary.
    candidateStore.removeAllOwnedCandidates()
}

struct AidenBotImagePlaygroundView: View {
    let identity: AidenBotImagePlaygroundIdentity
    let fallbackOverride: AidenBotImagePlaygroundFallbackReason?
    let candidateStore: AidenBotImagePlaygroundCandidateStore
    let onCandidateCopied: (URL) -> Void

    init(
        identity: AidenBotImagePlaygroundIdentity,
        fallbackOverride: AidenBotImagePlaygroundFallbackReason? = nil,
        candidateStore: AidenBotImagePlaygroundCandidateStore = .init(),
        onCandidateCopied: @escaping (URL) -> Void
    ) {
        self.identity = identity
        self.fallbackOverride = fallbackOverride
        self.candidateStore = candidateStore
        self.onCandidateCopied = onCandidateCopied
    }

    // `onCandidateCopied` transfers ownership of the copied URL. Its receiver
    // must remove it after replacement, upload completion, cancellation, or
    // teardown by calling `candidateStore.removeOwnedCandidate(at:)`.

    @ViewBuilder
    var body: some View {
        Group {
            if #available(iOS 18.1, *) {
                AidenBotSystemImagePlaygroundView(
                    identity: identity,
                    fallbackOverride: fallbackOverride,
                    candidateStore: candidateStore,
                    onCandidateCopied: onCandidateCopied
                )
            } else {
                AidenBotImagePlaygroundFallbackView(reason: fallbackOverride ?? .updateRequired)
            }
        }
        .task {
            // Retry launch cleanup when the editor becomes visible in case
            // protected temporary files were inaccessible while locked.
            candidateStore.removeAllOwnedCandidates()
        }
    }
}

@available(iOS 18.1, *)
private struct AidenBotSystemImagePlaygroundView: View {
    let identity: AidenBotImagePlaygroundIdentity
    let fallbackOverride: AidenBotImagePlaygroundFallbackReason?
    let candidateStore: AidenBotImagePlaygroundCandidateStore
    let onCandidateCopied: (URL) -> Void

    @Environment(\.supportsImagePlayground) private var supportsImagePlayground
    @State private var presentation = AidenBotImagePlaygroundPresentationState()

    private var isPresenting: Binding<Bool> {
        Binding(
            get: { presentation.phase == .presenting },
            set: { newValue in
                if !newValue, presentation.phase == .presenting {
                    presentation.cancel()
                }
            }
        )
    }

    private var concepts: [ImagePlaygroundConcept] {
        identity.conceptTexts.map(ImagePlaygroundConcept.text)
    }

    @ViewBuilder
    var body: some View {
        if let fallbackOverride {
            AidenBotImagePlaygroundFallbackView(reason: fallbackOverride)
        } else if #unavailable(iOS 18.4) {
            AidenBotImagePlaygroundFallbackView(reason: .updateRequired)
        } else if !supportsImagePlayground {
            AidenBotImagePlaygroundFallbackView(reason: .systemUnavailable)
        } else {
            configuredSheet
        }
    }

    @available(iOS 18.4, *)
    private var configuredSheet: some View {
        launcherContent
            .imagePlaygroundSheet(
                isPresented: isPresenting,
                concepts: concepts,
                onCompletion: captureCandidate,
                onCancellation: { presentation.cancel() }
            )
            .imagePlaygroundGenerationStyle(
                .illustration,
                in: [.animation, .illustration, .sketch]
            )
            .imagePlaygroundPersonalizationPolicy(.disabled)
    }

    private var launcherContent: some View {
        VStack(alignment: .leading, spacing: 12) {
            Label("Create with Apple Intelligence", systemImage: "apple.intelligence")
                .font(.headline)

            Text(Self.processingDisclosure)
                .font(.caption)
                .foregroundStyle(.secondary)

            if case .cancelled = presentation.phase {
                Label("No image was selected. Your semantic avatar is unchanged.", systemImage: "checkmark.circle")
                    .font(.caption)
                    .foregroundStyle(.secondary)
            } else if case .accepted = presentation.phase {
                Label("The selected image is ready to preview.", systemImage: "checkmark.circle.fill")
                    .font(.caption)
                    .foregroundStyle(.secondary)
            } else if case let .fallback(reason) = presentation.phase {
                AidenBotImagePlaygroundFallbackView(reason: reason)
            }

            Button {
                presentation.requestPresentation(systemAvailable: supportsImagePlayground)
            } label: {
                Label("Open Image Playground", systemImage: "photo.badge.plus")
            }
            .buttonStyle(.bordered)
            .disabled(presentation.phase == .presenting)
            .accessibilityHint("Opens Apple's system image creation sheet.")
        }
    }

    private func captureCandidate(_ temporaryURL: URL) {
        do {
            let copiedURL = try candidateStore.copyImmediately(fromSystemCompletionURL: temporaryURL)
            presentation.acceptCopiedCandidate()
            onCandidateCopied(copiedURL)
        } catch {
            presentation.failCandidateCopy()
        }
    }

    static let processingDisclosure = "Apple creates images in its system Image Playground. Processing is controlled by Apple and may use Private Cloud Compute. Aiden receives only the image you choose."
}

private struct AidenBotImagePlaygroundFallbackView: View {
    let reason: AidenBotImagePlaygroundFallbackReason

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            Label(reason.title, systemImage: "photo.badge.exclamationmark")
                .font(.headline)
            Text(reason.message)
                .font(.caption)
                .foregroundStyle(.secondary)
        }
        .accessibilityElement(children: .combine)
    }
}
